pages/models.py

```diff
@@ -30,12 +31,205 @@
+                'text_alignment': {'max_num': 1},
+            }, required=False))
+        ])),
+        ('divider', DividerBlock()),
+        ('html', blocks.RawHTMLBlock()),
+    ], use_json_field=True)
+
+    promote_image = models.ForeignKey(
+        'wagtailimages.Image',
+        null=True,
+        blank=False,
+        on_delete=models.SET_NULL,
+        related_name='+'
+    )
+
+    api_fields = [
+        APIField('layout', serializer=LayoutSnippetSerializer()),
+        APIField('body'),
+        APIField('slug'),
+        APIField('seo_title'),
+        APIField('search_description'),
+    ]
+
+    content_panels = [
+        TitleFieldPanel('title', help_text="For CMS use only. Use 'Promote' tab above to edit SEO information."),
+        FieldPanel('layout'),
+        FieldPanel('body'),
+    ]
+
+    promote_panels = [
+        FieldPanel('slug', widget=SlugInput),
+        FieldPanel('seo_title'),
+        FieldPanel('search_description'),
+        FieldPanel('promote_image')
+    ]
+
+    template = 'page.html'
+    preview_modes = []
+    max_count = 1
+    # TODO: we are allowing this to be built as a child of the homepage. Not ideal.
+    # Once the home page is released, use something to migrate homepage children to root page and remove this parent type.
+    parent_page_types = ['wagtailcore.Page', 'pages.HomePage']
+    subpage_types = ['pages.FlexPage']  # which might also require allowing all pages to be children.
+
+    def __str__(self):
+        return self.path
+
+    def get_url_parts(self, *args, **kwargs):
+        url_parts = super(RootPage, self).get_url_parts(*args, **kwargs)
+
+        if url_parts is None:
+            # in this case, the page doesn't have a well-defined URL in the first place -
+            # for example, it's been created at the top level of the page tree
+            # and hasn't been associated with a site record
+            return None
+
+        site_id, root_url, page_path = url_parts
+
+        # return '/' in place of the real page path for the root page
+        return site_id, root_url, '/'
+
+    def get_sitemap_urls(self, request=None):
+        return [
+            {
+                'location': '{}/'.format(Site.find_for_request(request).root_url),
+                'lastmod': (self.last_published_at or self.latest_revision_created_at),
+            }
+        ]
+
+# subclass of RootPage with a few overrides for subpages
+class FlexPage(RootPage):
+    parent_page_types = ['pages.RootPage']
+    template = 'page.html'
+
+    def get_url_parts(self, *args, **kwargs):
+        url_parts = super(FlexPage, self).get_url_parts(*args, **kwargs)
+
+        if url_parts is None:
+            return None
+
+        site_id, root_url, page_path = url_parts
+
+        return site_id, root_url, page_path
+
+    def get_sitemap_urls(self, request=None):
+        return [
+            {
+                'location': '{}/{}'.format(Site.find_for_request(request).root_url, self.slug),
```

Comment:
is this going to result in nested location paths if the page tree in the cms is nested? the assumption in routing is that all page slugs are at one level, not nested, but it looks here like you're combining levels

Reply:
this is only here to build the sitemap.xml

this particular block is for pages under the homepage (which should follow slug patterns, for e.g. /impact). The root_url is the environment name (set up in the CMS under Sites).